src/node.ts

```diff
@@ -0,0 +1,115 @@
+import {existsSync} from "node:fs";
+import {copyFile, readFile, writeFile} from "node:fs/promises";
+import {createRequire} from "node:module";
+import op from "node:path";
+import {extname, join} from "node:path/posix";
+import {pathToFileURL} from "node:url";
+import {nodeResolve} from "@rollup/plugin-node-resolve";
+import type {AstNode, OutputChunk, Plugin, ResolveIdResult} from "rollup";
+import {rollup} from "rollup";
+import esbuild from "rollup-plugin-esbuild";
+import {prepareOutput, toOsPath} from "./files.js";
+import type {ImportReference} from "./javascript/imports.js";
+import {isJavaScript, parseImports} from "./javascript/imports.js";
+import {parseNpmSpecifier} from "./npm.js";
+import {isPathImport} from "./path.js";
+import {faint} from "./tty.js";
+
+export async function resolveNodeImport(root: string, spec: string): Promise<string> {
+  return resolveNodeImportInternal(op.join(root, ".observablehq", "cache", "_node"), root, spec);
+}
+
+const bundlePromises = new Map<string, Promise<void>>();
+
+async function resolveNodeImportInternal(cacheRoot: string, packageRoot: string, spec: string): Promise<string> {
+  const {name, path = "."} = parseNpmSpecifier(spec);
+  const require = createRequire(pathToFileURL(op.join(packageRoot, "/")));
+  const pathResolution = require.resolve(spec);
+  let packageResolution = pathResolution;
+  do {
+    const p = op.dirname(packageResolution);
+    if (p === packageResolution) throw new Error(`unable to resolve package.json: ${spec}`);
+    packageResolution = p;
+  } while (!existsSync(op.join(packageResolution, "package.json")));
+  const {version} = JSON.parse(await readFile(op.join(packageResolution, "package.json"), "utf-8"));
```

Comment:
For posterity, here is an alternative approach to finding the `package.json` that I discovered:

https://github.com/nodejs/node/issues/33460#issuecomment-919184789

I guess there are some cases of a “nested” `package.json` that this doesn’t handle, and so their approach is to trim the relative path to the entry point from the fully-resolved path. I think our approach is fine for now, though. There’s also this:

https://github.com/sindresorhus/package-up

Reply:
I’ve switched this branch to use [pkg-dir](https://github.com/sindresorhus/pkg-dir).